fairscale/optim/oss.py

```diff
@@ -310,12 +297,116 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+
+        # - go through the per-shard states, which are all indexed locally
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        global_id = global_id_map[local_index_to_param_id[local_param_index]]
+                        state_dict["state"][global_id] = s["state"][local_param_index]
+
+        return state_dict
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
+
+        Arguments:
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
+        """
+
+        # NOTE: PyTorch 1.5 does not index linearly but with the id(params) at saving time
+        # we work around that here by using the fact that the params are ordered as in the param_groups
+
+        # Param index to param map
+        global_id_map = {i: p for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # Prune the state_dict from the states which this rank does not own, then normal base load
+        for i_param, key in enumerate(state_dict["state"].keys()):
+            # Check that this rank owns this param, if not remove from the state
+            param = global_id_map[i_param]
+            if self.param_to_rank[param] != self.rank:
+                state_dict["state"][key] = None
+
+        super().load_state_dict(state_dict)
+
+        # Set the sharded optimizer state.
+        # Keep the original type (not respected by PyTorch which casts to the model type)
+        for k, (_, v) in enumerate(state_dict["state"].items()):
+            if k in global_id_map:
+                param = global_id_map[k]
+
+                # Only add this state to the sharded optimizer if it owns this param
+                for pg in self.optim.param_groups:
+                    if id(param) in [id(p) for p in pg["params"]]:
+                        self.optim.state[param] = recursive_copy_to_device(v, non_blocking=True, device=param.device)
+
+        # Update the param_group keys
+        for fpg, pg in zip(state_dict["param_groups"], self.param_groups):
+            for key in fpg.keys():
+                if key != "params":
+                    pg[key] = fpg[key]
+
+        # Sync with the optimizer param groups
+        self._sync_param_groups(local_to_global=False)
```

Comment:
Might be simpler if there was just a function which synced between a src and dst param group
```suggestion
        self._sync_param_groups(self.param_groups, state_dict["param_groups"])
        self._sync_param_groups(self.optim.param_groups, self.param_groups)
```

Reply:
good point, looks a lot nicer indeed

Reply:
(the first line is flipped though)